src/main/java/com/shapesecurity/shift/es2017/template/Template.java

```diff
@@ -48,12 +48,32 @@ public Template(String source) {
 		}
 	}
 
+	public static Template fromModuleSource(String source) {
+		ParserWithLocation parserWithLocation = new ParserWithLocation();
+		try {
+			Program tree = parserWithLocation.parseModule(source);
+			ImmutableList<Template.NodeInfo> namePairs = findNodes(tree, parserWithLocation, parserWithLocation.getComments());
+			return new Template(tree, namePairs);
+		} catch (JsError e) {
+			throw new IllegalArgumentException("template failed to parse");
+		}
+	}
+
+	public static Template fromScriptSource(String source) {
+		return new Template(source);
+	}
+
+	public Template(Program tree, ImmutableList<NodeInfo> namePairs) {
```

Comment:
You should defer to this constructor in the string-typed constructor.

Reply:
Can't; java's dumb.